### ISSUE TITLE
[Bug]: postgres doesn't care what UID it runs in, but initdb does. user does not exist

### ISSUE THREAD

**Issue by wibed** [cloudnative-pg/cloudnative-pg#3331]
### Is there an existing issue already for this bug?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new bug.

### I have read the troubleshooting guide

- [X] I have read the troubleshooting guide and I think this is a new bug.

### I am running a supported version of CloudNativePG

- [X] I have read the troubleshooting guide and I think this is a new bug.

### Version

1.21.0

### What version of Kubernetes are you using?

1.28

### What is your Kubernetes environment?

Self-managed: k3s

### How did you install the operator?

Helm

### What happened?

i found a configuration item.

https://github.com/cloudnative-pg/cloudnative-pg/blob/06871b2735823111a3f78ccc57c384447ee1cde5/config/crd/bases/postgresql.cnpg.io_clusters.yaml#L2517

upon using postgresGID and postgresUID, the uid is not changed the user is:
```shell
# tl;dr: "logger":"initdb","msg":"initdb: could not look up effective user ID 501: user does not exist\n",
# full log below:
{"level":"error","ts":"2023-11-08T07:54:09Z","msg":"Error while bootstrapping data directory","logging_pod":"cluster0-1-initdb","error":"error while creating the PostgreSQL instance: exit status 1","stacktrace":"github.com/cloudnative-pg/cloudnative-pg/pkg/management/log.(*logger).Error\n\tpkg/management/log/log.go:128\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/log.Error\n\tpkg/management/log/log.go:166\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/instance/initdb.initSubCommand\n\tinternal/cmd/manager/instance/initdb/cmd.go:151\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/instance/initdb.NewCmd.func2\n\tinternal/cmd/manager/instance/initdb/cmd.go:104\ngithub.com/spf13/cobra.(*Command).execute\n\tpkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tpkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068\ngithub.com/spf13/cobra.(*Command).Execute\n\tpkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992\nmain.main\n\tcmd/manager/main.go:64\nruntime.main\n\t/opt/hostedtoolcache/go/1.21.3/x64/src/runtime/proc.go:267"}
Error: error while creating the PostgreSQL instance: exit status 1
```
 

the doc states
```
--username=username 

Selects the user name of the database superuser. 
This defaults to the name of the effective user running initdb. 
It is really not important what the superuser's name is, 
but one might choose to keep the customary name postgres, 
even if the operating system user's name is different.
```

therefor to set uid/gid expects the system to create the user before starting the process. which it does not

relevant excerpt from: https://hub.docker.com/_/postgres
```yaml
The main caveat to note is that postgres doesn't care what UID it runs as (as long as the owner of /var/lib/postgresql/data matches), but initdb does care (and needs the user to exist in /etc/passwd):
```

the easiest way around this it to include a chmod operation before the pod starts.

bitnami has an implementation of such a process.
https://github.com/bitnami/charts/blob/main/bitnami/postgresql/templates/primary/statefulset.yaml

### what do i expect
i would like for the instance manager to create the user if its uid is missing.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by wibed** [comment]
i found the exec command

https://github.com/cloudnative-pg/cloudnative-pg/blob/8cceb710d459bf93da94a5ca115c63e2908afdc5/pkg/utils/exec.go#L42


gonna look at the bootstrap

**Comment by ndrwstn** [comment]
It seems to be that ignoring `containerSecurityContext` is the problem, unless I don't understand the problem. Why doesn't  `runAsUser`, `runAsGroup` (and maybe `fsGroup`) override `postgresUID` and `postgresGID`?

**Comment by wibed** [comment]
@ndrwstn 
well its in the title.

pg wont care if a user exists on the system but initdb, initializing the pg db does.
so if you provide a uid/gid which does not exists your pods wont start up because of it.

**Comment by ndrwstn** [comment]
I don't understand your response,then Isn't the point of `containerSecurityContext` to provide the user/group that the container is supposed to run as. I guess there is a scenario where someone is assigning that to a user that doesn't exist, but why? And if initdb respects the security context (and we trust the end-user not to put nonsense into security context), doesn't that solve the problem?

I'm just looking for an alternative answer as it didn't seem like they're going to accept your PR due to the need for root access.

**Comment by wibed** [comment]
running the container with the gid/uid from nobody as sensible default might be the safest bet security wise and a valid reason. 

iirc the securityContext in itself does not change the underlying system in itself. it just runs the container along the asigned gid/uid.

**Comment by ndrwstn** [comment]
I hadn't thought of that, but it's a good point.

**Comment by mnencia** [comment]
There seems to be a misunderstanding regarding the CNPG operator. It assumes that the container images it runs have a `postgres` user with UID 26 (which is the default in RHEL systems). However, if your image requires a different value, you can change the UIDs by using the `postgresUID` and `postgresGID` fields. It is the responsibility of the image creator to ensure that the user exists.

I would like to highlight that `initdb` is not the only tool that requires an environment to have a line in `/etc/passwd` for the current UID. There are many other tools and extensions that also require this, so it is not safe to assume that a tool will work correctly in a misconfigured container.

It is the responsibility of the person who builds the container image to include the postgres user inside the Postgres container. If the environment changes the user to something different, such as OpenShift's restricted SCC, then it is up to the environment to make sure that the user exists within the /etc/password file, and our operator supports it.

**Comment by wibed** [comment]
@mnencia given id like to add such a user beforehand. How would i approach this? 
to shuffle the responsibility onto the user is easy, but if its not possible to solve it might be the wrong thing to do.

**Comment by sxd** [comment]
Hello @wibed and @ndrwstn 

From a security point of view, this is not plausible since there's system that will not rely on the user provided and will 
change it applying different security policies, so this will be blocking everything that may use a different user.

Following on the security side, creating a UID or a GID requires super user privileges, the images are, and should always be, by default, run as non root, having a root user inside a PostgreSQL image, is, in all the possible meanings a really bad idea, and should never be the use case, now, going back to the patch, if that's required, this it's a fully no go for this.

Related to the image, presumes that a couple of commands are there, well, the major problem here will be with a PostgreSQL image using alpine or ubi, even worst, what if we use a distroless image? this for sure will not work.

I think it's clear here that the problem shouldn't be handled at runtime level but when the image is created, but moreover, here is missing a way to reproduce the issue, I don't fully understand because we can't reproduce, can you add the steps to reproduce here so we can see more clearly how you reached this point?

Regards,
